src/damage.cpp

```diff
@@ -497,45 +497,56 @@ damage_instance load_damage_instance_inherit( const JsonArray &jarr, const damag
     return di;
 }
 
-std::map<damage_type, float> load_damage_map( const JsonObject &jo )
-{
-    std::map<damage_type, float> ret;
-    std::optional<float> init_val = jo.has_float( "all" ) ?
-                                    jo.get_float( "all", 0.0f ) :
-                                    std::optional<float>();
-
-    auto load_if_present = [&ret, &jo]( const std::string & name, damage_type dt,
-    std::optional<float> fallback ) {
-        if( jo.has_float( name ) ) {
-            float val = jo.get_float( name );
-            ret[dt] = val;
-        } else if( fallback ) {
-            ret[dt] = *fallback;
+namespace
+{
+
+struct DamageMapping { std::string name; damage_type type; bool physical; };
+
+constexpr auto damage_mappings = std::array<DamageMapping, 12>
+{{
+        { .name = "bash",       .type = DT_BASH,       .physical = true  },
+        { .name = "cut",        .type = DT_CUT,        .physical = true  },
+        { .name = "stab",       .type = DT_STAB,       .physical = true  },
+        { .name = "bullet",     .type = DT_BULLET,     .physical = true  },
+        { .name = "biological", .type = DT_BIOLOGICAL, .physical = false },
+        { .name = "acid",       .type = DT_ACID,       .physical = false },
+        { .name = "heat",       .type = DT_HEAT,       .physical = false },
+        { .name = "cold",       .type = DT_COLD,       .physical = false },
+        { .name = "dark",       .type = DT_DARK,       .physical = false },
+        { .name = "light",      .type = DT_LIGHT,      .physical = false },
+        { .name = "psi",        .type = DT_PSI,        .physical = false },
+        { .name = "electric",   .type = DT_ELECTRIC,   .physical = false }
+    }
+};
+
+auto get_float_optional( const JsonObject &jo, const std::string &name ) -> std::optional<float>
+{
+    return jo.has_member( name ) ? std::make_optional( jo.get_float( name, 0.0f ) ) : std::nullopt;
+}
```

Comment:
If a fallback value is added we can slightly optimize the loop where the generated values are used
```suggestion
auto get_float_optional( const JsonObject &jo, const std::string &name, const std::optional<float> fallback ) -> std::optional<float>
{
    return jo.has_member( name ) ? std::make_optional( jo.get_float( name, 0.0f ) ) : fallback;
}
```